<launch>
  <arg name="use_recovery_behaviors" default="true"/>
  
  <arg name="egocircle" default="false"/>
  
  <arg name="converter" default="false" if="$(arg egocircle)"/>
  <arg name="converter" default="true" unless="$(arg egocircle)"/>
  
  <arg name="simple_exploration" default="false"/>
  
  <!--<include file="$(find turtlebot_navigation)/launch/includes/velocity_smoother.launch.xml"/>-->
  <include file="$(find closedloop_nav_slam)/launch/nav/velocity_smoother.launch"/>

  <node name="egocircle_node" pkg="egocircle" type="ego_circle" if="$(arg egocircle)"/>
  
  <arg name="global_planning_freq" default="$(optenv GM_PARAM_GLOBAL_PLANNING_FREQ 1)"/>
  <arg name="controller_freq" default="$(optenv GM_PARAM_CONTROLLER_FREQ 5)"/>

  <arg name="debug" default="false"/>

  <arg name="debug_prefix" value="" unless="$(arg debug)"/> <!-- valgrind - -tool=callgrind -->
  <arg name="debug_prefix" value="xterm -e gdb -ex run --args" if="$(arg debug)"/>

  <arg name="goal_reached_thresh" default="0.2"/>
  <arg name="goal_reached_orient_thresh" default="1.58"/>

  <!-- xterm -e gdb -ex run - -args -->
  
  
  <!--<env name="ROSCONSOLE_CONFIG_FILE" value="$(find nav_configs)/config/custom_teb_rosconsole.conf"/>-->

  <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen" clear_params="true" launch-prefix="$(arg debug_prefix)">
      <rosparam file="$(find closedloop_nav_slam)/configs/params/nav/common_costmap_params.yaml" command="load" ns="global_costmap"/>
      <rosparam file="$(find closedloop_nav_slam)/configs/params/nav/common_costmap_params.yaml" command="load" ns="local_costmap"/>
      <rosparam file="$(find closedloop_nav_slam)/configs/params/nav/local_costmap_params.yaml" command="load"/>
      <rosparam file="$(find closedloop_nav_slam)/configs/params/nav/global_costmap_params.yaml" command="load"/>

      <rosparam file="$(find closedloop_nav_slam)/configs/params/nav/local_planner_common_params.yaml" command="load" ns="TebLocalPlannerROS"/>
      <rosparam file="$(find closedloop_nav_slam)/configs/params/nav/move_base_params.yaml" command="load"/>
      <rosparam file="$(find closedloop_nav_slam)/configs/params/nav/global_planner_params.yaml" command="load" ns="GlobalPlanner"/>

      <param name="recovery_behavior_enabled" value="false" unless="$(arg use_recovery_behaviors)"/>

      <param name="controller_frequency" value="$(arg controller_freq)" />
      <param name="planner_frequency" value="$(arg global_planning_freq)"/>

      <remap from="cmd_vel" to="navigation_velocity_smoother/raw_cmd_vel"/>


      <!-- Approach-specific parameters/remappings/etc -->
      <param name="base_local_planner" value="teb_local_planner/TebLocalPlannerROS" />

      <rosparam file="$(find closedloop_nav_slam)/configs/params/nav/teb_local_planner_params.yaml" command="load"/>
      <rosparam file="$(find closedloop_nav_slam)/configs/params/nav/teb_costmap_converter_params.yaml" command="load" if="$(arg converter)"/>

      <param name="TebLocalPlannerROS/include_costmap_obstacles" value="false" if="$(arg egocircle)"/>
      <param name="TebLocalPlannerROS/include_egocircle_obstacles" value="$(arg egocircle)"/>
      <param name="TebLocalPlannerROS/gap_h_signature" value="$(arg egocircle)"/>
      <param name="TebLocalPlannerROS/gap_exploration" value="$(arg egocircle)"/>
      <param name="TebLocalPlannerROS/egocircle_feasibility" value="$(arg egocircle)"/>
      <param name="TebLocalPlannerROS/simple_exploration" value="$(arg simple_exploration)"/>
      <param name="TebLocalPlannerROS/weight_gap" value="0"/>
      <param name="TebLocalPlannerROS/gap_boundary_exponent" value="2" unless="$(arg egocircle)"/>
      <param name="TebLocalPlannerROS/feasibility_check_no_tebs" value="1" if="$(arg egocircle)"/>
      <param name="TebLocalPlannerROS/gap_theta_start" value="3.15"/>
      
      <param name="TebLocalPlannerROS/max_global_plan_lookahead_dist" value="0"/>
      <param name="TebLocalPlannerROS/xy_goal_tolerance" value="$(arg goal_reached_thresh)"/>
      <param name="TebLocalPlannerROS/yaw_goal_tolerance" value="$(arg goal_reached_orient_thresh)"/>
      
      <!-- Disable the local costmap observation sources if they will not be used -->
      <param name="local_costmap/obstacle_layer/observation_sources" value="false" if="$(arg egocircle)"/>
  </node>





</launch>
